### PR TITLE
Feat/custom strategy screen

### PR DESCRIPTION
**Pull request by ykhedher**
https://www.loom.com/share/26c633f8006b4e14b657aad37eb28554

### CODE REVIEW

**Comment by ykhedher** [comment]
[Create custom strategies screen](https://trello.com/c/VE4PHYeG/882-create-custom-strategies-screen)

File: src/component/strategies/StrategyForm/StrategyForm.tsx [hunk 1]
```diff
@@ -1,194 +1,111 @@
...
-import { formatUnknownError } from '../../../utils/format-unknown-error';
+import { StrategyParameters } from './StrategyParameters/StrategyParameters';
+import { ICustomStrategyParameter } from 'interfaces/strategy';
 
-interface ICustomStrategyParams {
-    name?: string;
-    type?: string;
-    description?: string;
-    required?: boolean;
+interface IStrategyForm {
```
Comment (olav) [inline src/component/strategies/StrategyForm/StrategyForm.tsx:9]: `IStrategyFormProps`

File: src/component/strategies/StrategyForm/StrategyForm.tsx [hunk 2]
```diff
@@ -1,194 +1,111 @@
...
+    strategyName: string;
+    strategyDesc: string;
+    params: ICustomStrategyParameter[];
+    setStrategyName: React.Dispatch<React.SetStateAction<string>>;
+    setStrategyDesc: React.Dispatch<React.SetStateAction<string>>;
+    setParams: React.Dispatch<React.SetStateAction<ICustomStrategyParameter[]>>;
+    handleSubmit: (e: any) => void;
```
Comment (olav) [inline src/component/strategies/StrategyForm/StrategyForm.tsx:16]: Could use a stricter type:

```
handleSubmit: (e: React.FormEvent<HTMLFormElement>) => void;
```

File: src/component/strategies/StrategyForm/StrategyForm.tsx [hunk 3]
```diff
@@ -1,194 +1,111 @@
...
+    handleCancel: () => void;
+    errors: { [key: string]: string };
+    mode: string;
```
Comment (olav) [inline src/component/strategies/StrategyForm/StrategyForm.tsx:19]: Could be stricter here as well:

```
mode: 'create' | 'edit'
```

File: src/component/strategies/StrategyForm/StrategyForm.tsx [hunk 4]
```diff
@@ -1,194 +1,111 @@
...
+    clearErrors: () => void;
+    setErrors: React.Dispatch<React.SetStateAction<Object>>;
```
Comment (olav) [inline src/component/strategies/StrategyForm/StrategyForm.tsx:21]: Object is tricky. Maybe:

```
setErrors: React.Dispatch<React.SetStateAction<Record<string, string>>>;
```

File: src/component/strategies/StrategyForm/StrategyForm.tsx [hunk 5]
```diff
@@ -1,194 +1,111 @@
...
-                            onCancel={handleCancel}
-                        />
-                    }
-                />
-            </form>
-        </PageContent>
+            </div>
+            <div className={styles.buttonContainer}>
+                {children}
+                <Button onClick={handleCancel} className={styles.cancelButton}>
```
Comment (olav) [inline src/component/strategies/StrategyForm/StrategyForm.tsx:105]: Should add a `type="button"` here, to make sure this button won't submit the form.

File: src/component/strategies/StrategyForm/StrategyParameters/StrategyParameter/StrategyParameter.tsx [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+export const StrategyParameter = ({
+    set,
+    input,
+    index,
+    params,
+    setParams,
+    errors,
+}: IStrategyParameterProps) => {
+    const styles = useStyles();
+    const handleTypeChange = (event: { target: { value: any } }) => {
```
Comment (olav) [inline src/component/strategies/StrategyForm/StrategyParameters/StrategyParameter/StrategyParameter.tsx:60]: We have a type that might work here:

```
const handleTypeChange: OnGeneralSelectChange = (event) => {
```

File: src/component/strategies/StrategyForm/StrategyParameters/StrategyParameter/StrategyParameter.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                    className={styles.name}
+                    error={Boolean(errors?.[`paramName${index}`])}
+                    errorText={errors?.[`paramName${index}`]}
+                />
+                <IconButton
+                    onClick={() => {
+                        setParams(params.filter((e, i) => i !== index));
+                    }}
+                >
+                    <Delete />
```
Comment (olav) [inline src/component/strategies/StrategyForm/StrategyParameters/StrategyParameter/StrategyParameter.tsx:95]:
```
<Delete titleAccess="Delete" />
```

File: src/component/strategies/StrategyForm/StrategyParameters/StrategyParameters.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+export const StrategyParameters = ({
+    input = [],
+    count = 0,
+    updateParameter,
+    setParams,
+    errors,
+}: IStrategyParametersProps) => (
+    <div>
+        {gerArrayWithEntries(count).map((v, i) => (
```
Comment (olav) [inline src/component/strategies/StrategyForm/StrategyParameters/StrategyParameters.tsx:25]: Maybe we can use `input.map((item, i) => ...` insteead of the count prop.

File: src/component/strategies/StrategyView/StrategyDetails/StrategyDetails.tsx [hunk 9]
```diff
@@ -11,14 +11,16 @@ import { AppsLinkList } from '../../../common';
...
+    IStrategy,
+} from '../../../../interfaces/strategy';
 import { IApplication } from '../../../../interfaces/application';
-import { IFeatureToggle } from '../../../../interfaces/featureToggle';
 
 interface IStrategyDetailsProps {
     strategy: IStrategy;
     applications: IApplication[];
-    toggles: IFeatureToggle[];
+    toggles: [];
```
Comment (olav) [inline src/component/strategies/StrategyView/StrategyDetails/StrategyDetails.tsx:23]: :eyes:

File: src/component/strategies/StrategyView/StrategyView.tsx [hunk 10]
```diff
@@ -1,75 +1,64 @@
...
+                        <ConditionallyRender
+                            condition={strategy.editable}
+                            show={
+                                <PermissionIconButton
+                                    permission={UPDATE_STRATEGY}
+                                    tooltip={'Edit strategy'}
+                                    data-loading
+                                    onClick={handleEdit}
+                                >
+                                    <Edit />
```
Comment (olav) [inline src/component/strategies/StrategyView/StrategyView.tsx:48]:
```
<Edit titleAccess="Edit strategy" />
```

File: src/component/strategies/hooks/useStrategyForm.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        setStrategyName(initialStrategyName);
+    }, [initialStrategyName]);
+
+    useEffect(() => {
+        setStrategyDesc(initialStrategyDesc);
+    }, [initialStrategyDesc]);
+
+    useEffect(() => {
+        setParams(initialParams);
+    }, [JSON.stringify(initialParams)]);
```
Comment (olav) [inline src/component/strategies/hooks/useStrategyForm.ts:26]: Will the initial values change after the first call? If not, these effects cound be removed.

File: src/component/strategies/hooks/useStrategyForm.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+    const getStrategyPayload = () => {
+        return {
+            name: strategyName,
+            description: strategyDesc,
+            parameters: params,
+        };
+    };
+
+    const validateStrategyName = async () => {
```
Comment (olav) [inline src/component/strategies/hooks/useStrategyForm.ts:36]: Should remove `async`.

File: src/hooks/api/getters/useStrategy/useStrategy.ts [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import handleErrorResponses from '../httpErrorResponseHandler';
+
+const useStrategy = (strategyName: string, options: SWRConfiguration = {}) => {
+    const STRATEGY_CACHE_KEY = `api/admin/strategies/${strategyName}`;
+    const path = formatApiPath(STRATEGY_CACHE_KEY);
+
+    const fetcher = () => {
+        return fetch(path, {
+            method: 'GET',
+            credentials: 'include',
```
Comment (olav) [inline src/hooks/api/getters/useStrategy/useStrategy.ts:14]: Can remove these options. GET is the default method, and we're not doing cross-origin requests AFAIK.

File: src/hooks/api/getters/useStrategy/useStrategy.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+            .then(handleErrorResponses(`${strategyName} strategy`))
+            .then(res => res.json());
+    };
+
+    const { data, error } = useSWR<{ strategy: IStrategy }>(
+        STRATEGY_CACHE_KEY,
+        fetcher,
+        options
+    );
+    const [loading, setLoading] = useState(!error && !data);
```
Comment (olav) [inline src/hooks/api/getters/useStrategy/useStrategy.ts:25]: Don't need state for `loading`, could return `loading: !error && !data` instead.

File: src/component/strategies/hooks/useStrategyForm.ts [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+            }));
+            return false;
+        }
+        return true;
+    };
+
+    const validateParams = () => {
+        let res = true;
+        //@ts-expect-error
+        for (const [index, p] of params.entries()) {
```
Comment (olav) [inline src/component/strategies/hooks/useStrategyForm.ts:54]: How about one of these to get rid of the error:

```
for (const [index, p] of Object.entries(params)) { // ...

params.forEach((p, index) => // ...
```

**Review by olav — CHANGES_REQUESTED** [review]

File: src/component/strategies/hooks/useStrategyForm.ts (hunk 11 above)
Comment (ykhedher) [reply src/component/strategies/hooks/useStrategyForm.ts:26]: These are used mainly for the Edit component, sometimes the response comes a bit late so we need a useEffect to update them.

File: src/component/strategies/StrategyForm/StrategyParameters/StrategyParameters.tsx (hunk 8 above)
Comment (FredrikOseberg) [reply src/component/strategies/StrategyForm/StrategyParameters/StrategyParameters.tsx:25]: Let's also use proper parameter names here `value / index`

File: src/component/strategies/StrategiesList/StrategiesList.tsx [hunk 16]
```diff
@@ -198,9 +199,37 @@ export const StrategiesList = () => {
         />
     );
 
-    const deleteButton = (strategy: IStrategy) => (
+    const editButton = (strategy: ICustomStrategyPayload) => (
         <ConditionallyRender
-            condition={strategy.editable}
+            //@ts-expect-error
```
Comment (FredrikOseberg) [inline src/component/strategies/StrategiesList/StrategiesList.tsx:204]: We should not add new // @ts-expect-error. We should fix the error instead.

File: src/component/strategies/StrategiesList/StrategiesList.tsx [hunk 17]
```diff
@@ -198,9 +199,37 @@ export const StrategiesList = () => {
...
+                        </IconButton>
+                    </div>
+                </Tooltip>
+            }
+        />
+    );
+
+    const deleteButton = (strategy: ICustomStrategyPayload) => (
+        <ConditionallyRender
+            //@ts-expect-error
```
Comment (FredrikOseberg) [inline src/component/strategies/StrategiesList/StrategiesList.tsx:231]: No new // @ts-expect-error

File: src/component/strategies/StrategyForm/StrategyParameters/StrategyParameter/StrategyParameter.tsx (hunk 6 above)
Comment (FredrikOseberg) [reply src/component/strategies/StrategyForm/StrategyParameters/StrategyParameter/StrategyParameter.tsx:60]: Implement this.

File: src/component/strategies/StrategyForm/StrategyParameters/StrategyParameter/StrategyParameter.tsx [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+                multiline
+                label={`Parameter name ${index + 1} description`}
+                onChange={({ target }) => set({ description: target.value })}
+                value={input.description}
+                className={styles.description}
+            />
+            <FormControlLabel
+                control={
+                    <Checkbox
+                        checked={!!input.required}
```
Comment (FredrikOseberg) [inline src/component/strategies/StrategyForm/StrategyParameters/StrategyParameter/StrategyParameter.tsx:121]: Boolean(input.required)

File: src/component/strategies/StrategyView/StrategyView.tsx [hunk 19]
```diff
@@ -1,77 +1,65 @@
...
-                                        strategy={strategy}
-                                        toggles={toggles}
-                                        applications={applications}
-                                    />
-                                </div>
-                            </section>
-                        }
+                    <StrategyDetails
+                        strategy={strategy}
+                        //@ts-expect-error
```
Comment (FredrikOseberg) [inline src/component/strategies/StrategyView/StrategyView.tsx:60]: Fix

**Review by FredrikOseberg — CHANGES_REQUESTED** [review]
Almost ready, but still a few things to fix up.

File: src/component/strategies/StrategyView/StrategyView.tsx [hunk 20]
```diff
@@ -1,77 +1,64 @@
...
-    const { strategyName } = useParams<{ strategyName: string }>();
+    const { name } = useParams<{ name: string }>();
     const { strategies } = useStrategies();
     const { features } = useFeatures();
     const { applications } = useApplications();
+    const history = useHistory();
 
     const toggles = features.filter(toggle => {
-        return toggle?.strategies?.find(s => s.name === strategyName);
+        return toggle?.strategies?.find(s => s.name === name);
```
Comment (FredrikOseberg) [inline src/component/strategies/StrategyView/StrategyView.tsx:22]:
```suggestion
        return toggle?.strategies?.find(strategy => strategy.name === name);
```

File: src/component/strategies/StrategyView/StrategyView.tsx [hunk 21]
```diff
@@ -1,77 +1,64 @@
...
     });
 
-    const strategy = strategies.find(n => n.name === strategyName);
+    const strategy = strategies.find(n => n.name === name);
```
Comment (FredrikOseberg) [inline src/component/strategies/StrategyView/StrategyView.tsx:25]:
```suggestion
    const strategy = strategies.find(strategy => strategy.name === name);
```

**Review by FredrikOseberg — APPROVED** [review]
LGTM

**Pull request merged**